commandapi-core/src/main/java/dev/jorel/commandapi/AbstractCommandAPICommand.java

```diff
@@ -212,29 +247,69 @@ public void register() {
 			CommandAPI.logWarning("Command /" + meta.commandName + " is being registered after the server had loaded. Undefined behavior ahead!");
 		}
 		try {
-			Argument[] argumentsArr = (Argument[]) (args == null ? new AbstractArgument[0] : args.toArray(AbstractArgument[]::new));
+			Argument[] argumentsArray = (Argument[]) (arguments == null ? new AbstractArgument[0] : arguments.toArray(AbstractArgument[]::new));
 
 			// Check IGreedyArgument constraints
-			for (int i = 0, numGreedyArgs = 0; i < argumentsArr.length; i++) {
-				if (argumentsArr[i] instanceof IGreedyArgument) {
-					if (++numGreedyArgs > 1 || i != argumentsArr.length - 1) {
-						throw new GreedyArgumentException(argumentsArr);
+			for (int i = 0, numGreedyArgs = 0; i < argumentsArray.length; i++) {
+				if (argumentsArray[i] instanceof IGreedyArgument) {
+					if (++numGreedyArgs > 1 || i != argumentsArray.length - 1) {
+						throw new GreedyArgumentException(argumentsArray);
 					}
 				}
 			}
 
+			// Check optional argument constraints
+			// They can only be at the end, no required argument can follow an optional argument
+			int firstOptionalArgumentIndex = -1;
+			for (int i = 0, optionalArgumentIndex = -1; i < argumentsArray.length; i++) {
+				if (argumentsArray[i].isOptional()) {
+					firstOptionalArgumentIndex = i;
+					optionalArgumentIndex = i;
+					continue;
+				}
+				// Argument is not optional
+				if (optionalArgumentIndex != -1) {
+					throw new OptionalArgumentException();
+				}
+			}
+
 			// Assign the command's permissions to arguments if the arguments don't already
 			// have one
-			for (Argument argument : argumentsArr) {
+			for (Argument argument : argumentsArray) {
 				if (argument.getArgumentPermission() == null) {
 					argument.withPermission(meta.permission);
 				}
 			}
 
+			// Create a List of arrays that hold arguments to register optional arguments
+			// if optional arguments have been found
+			List<Argument[]> argumentsToRegister = new ArrayList<>();
+			if (firstOptionalArgumentIndex != -1) {
+				for (int i = 0; i <= argumentsArray.length; i++) {
+					if (i == firstOptionalArgumentIndex - 1) {
+						Argument[] requiredArguments = (Argument[]) new AbstractArgument[i];
+						System.arraycopy(argumentsArray, 0, requiredArguments, 0, i);
+						argumentsToRegister.add(requiredArguments);
+						continue;
+					}
+					if (i >= firstOptionalArgumentIndex) {
```

Comment:
This could be an `else if` statement instead of requiring a `continue` statement. (See other review comment)